### PR TITLE
Athens: encode mods + fix handler bugs + stream mongo

### PR DESCRIPTION
**Pull request by marwan-at-work**
Fixes https://github.com/gomods/athens/issues/358, https://github.com/gomods/athens/issues/357, and half of https://github.com/gomods/athens/issues/356

Also some handler clean ups where we were calling WriteHeader twice.

### CODE REVIEW

File: pkg/paths/decode.go [hunk 1]
```diff
@@ -0,0 +1,301 @@
...
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+// Ripped from cmd/go
+
+// DecodePath returns the module path of the given safe encoding.
+// It fails if the encoding is invalid or encodes an invalid path.
+func DecodePath(encoding string) (path string, err error) {
```
Comment (michalpristas) [inline pkg/paths/decode.go]: please include table test with good/bad inputs, also serves as docs/examples

File: pkg/download/version_module.go [hunk 2]
```diff
@@ -23,18 +21,12 @@ func VersionModuleHandler(dp Protocol, lggr *log.Logger, eng *render.Engine) buf
...
-			err = errors.E(op, errors.M(mod), errors.V(ver), err)
-			status = http.StatusInternalServerError
-			lggr.SystemErr(err)
-		}
 
-		c.Response().WriteHeader(status)
-		return nil
+		// Calling c.Response().Write will write the header directly
+		// and we would get a 0 status in the buffalo logs.
+		return c.Render(200, eng.String(string(verInfo.Mod)))
```
Comment (arschles) [inline pkg/download/version_module.go]: 👍

File: pkg/download/version_zip.go [hunk 3]
```diff
@@ -25,17 +24,18 @@ func VersionZipHandler(dp Protocol, lggr *log.Logger, eng *render.Engine) buffal
...
+		// Calling c.Response().Write will write the header directly
+		// and we would get a 0 status in the buffalo logs.
+		c.Render(200, nil)
 		_, err = io.Copy(c.Response(), verInfo.Zip)
 		if err != nil {
-			status = http.StatusInternalServerError
 			lggr.SystemErr(errors.E(op, errors.M(mod), errors.V(ver), err))
 		}
 
-		c.Response().WriteHeader(status)
```
Comment (arschles) [inline pkg/download/version_zip.go]: is it cool with you to not send an error back down to the client in this case? and are we just unable to do it b/c of the buffalo bug?

File: pkg/paths/path.go [hunk 4]
```diff
@@ -31,9 +31,9 @@ func GetAllParams(c buffalo.Context) (*AllPathParams, error) {
 	if err != nil {
 		return nil, errors.E(op, err)
 	}
-	version := c.Param("version")
-	if version == "" {
-		return nil, errors.E(op, "missing version paramater")
+	version, err := DecodeVersion(c.Param("version"))
```
Comment (arschles) [inline pkg/paths/path.go]: can you remind me why we need `DecodeVersion`? and `DecodePath`? it seems like a lot of duplicated code that would be _amazing_ if we could avoid, so I'm wondering if there's any way around it

File: pkg/storage/mongo/getter.go [hunk 5]
```diff
@@ -1,18 +1,18 @@
...
 
 	"github.com/globalsign/mgo/bson"
+	"github.com/gomods/athens/pkg/errors"
 	"github.com/gomods/athens/pkg/storage"
 	opentracing "github.com/opentracing/opentracing-go"
 )
 
 // Get a specific version of a module
 func (s *ModuleStore) Get(module, vsn string) (*storage.Version, error) {
+	const op errors.Op = "mongo.Get"
```
Comment (arschles) [inline pkg/storage/mongo/getter.go]: @marwan-at-work I'm really liking this standardized error handling. just thought you should know - not really related to this PR

🎉

File: pkg/download/version_zip.go (hunk 3 above)
Comment (marwan-at-work) [reply pkg/download/version_zip.go]: @arschles Yes. The only possible error is that if the connection dropped or we went some falsy data, which in both cases the client will figure out. Plus, per the comment I put above, it's really our only option :)

File: pkg/paths/decode.go (hunk 1 above)
Comment (marwan-at-work) [reply pkg/paths/decode.go]: will do

File: pkg/paths/path.go (hunk 4 above)
Comment (marwan-at-work) [reply pkg/paths/path.go]: @arschles find the reason here :) https://github.com/gomods/athens/issues/356

File: pkg/storage/mongo/getter.go (hunk 5 above)
Comment (marwan-at-work) [reply pkg/storage/mongo/getter.go]: @arschles I appreciate you 🕺

**Comment by marwan-at-work** [comment]
@arschles yes, I can split the PR. The difference between GridFS and loading the whole zip into memory doesn't need benchmarking, it's visible to the naked eye lol

**Comment by arschles** [comment]
Fair enough! Thanks for splitting up

**Pull request closed without merging**